### PR TITLE
Remove ambiguous `padding_mask` and instead use a 2D->4D Attn Mask Mapper

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

For models that have Flash Attention 2 (FA2) implemented we currently pass both `padding_mask` and `attention_mask` to the respective vanilla attention class, *e.g.* `LlamaAttention` and to the FA2 class, *e.g.* `LlamaFlashAttention2`.

**However**, `padding_mask` is **not** used for `LlamaAttention` and `attention_mask` is not used for `LlamaFlashAttention2`. Conceptually the two masks are the same, only that `attention_mask` is a 4D mask while `padding_mask` is a 2D mask.

Passing around both masks and having both masks as concepts in our codebase is ambiguous and hurts readability. In this PR, I propose to remove the concept of `padding_mask` completely and instead just pass either a 2D or 4D `attention_mask` depending on whether we use FA2 or not.

**Note**: An additional benefit of this PR is that it will improve the performance when using FA2 as we will not create a 4D attention mask anymore.

**Benchmarks**:

The following script was used to benchmark the effect this mask implementation has on forward and generate.

```py
#!/usr/bin/env python3
from transformers import AutoTokenizer, AutoModelForCausalLM
import time
import torch

DEVICE = "cuda:1"

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf")
model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", torch_dtype=torch.float16, low_cpu_mem_usage=True)
model.to(DEVICE)


# forward
print("Forward benchmarks")
print(50 * "=")

for batch_size in (1, 4, 16):
    for input_seq in (4, 16, 256):
        input_ids = torch.ones((batch_size, input_seq), dtype=torch.long, device=DEVICE)
        attention_mask = torch.ones_like(input_ids)
        attention_mask[0, 3] = 0

        times = []
        for _ in range(3):
            start_time = time.time()
            with torch.no_grad():
                logits = model(input_ids=input_ids, attention_mask=attention_mask).logits
            times.append(time.time() - start_time)

        result = min(times)

        print(f"Forward bsz={batch_size}, input_seq={input_seq}: {result}")


# generate
print("Generate benchmarks")
print(50 * "=")

for batch_size in (1, 16):
    for input_seq in (4, 256):
        input_ids = torch.ones((batch_size, input_seq), dtype=torch.long, device=DEVICE)
        attention_mask = torch.ones_like(input_ids)
        attention_mask[0, 3] = 0

        times = []
        for _ in range(3):
            start_time = time.time()
            out = model.generate(input_ids=input_ids, max_new_tokens=256, do_sample=False)
            times.append(time.time() - start_time)

        result = min(times)

        print(f"Generate bsz={batch_size}, input_seq={input_seq}: {result}")
```

**This PR:**
```
Forward benchmarks
==================================================
Forward bsz=1, input_seq=4: 0.012479066848754883
Forward bsz=1, input_seq=16: 0.011297464370727539
Forward bsz=1, input_seq=256: 0.01240849494934082
Forward bsz=4, input_seq=4: 0.011190414428710938
Forward bsz=4, input_seq=16: 0.013025283813476562
Forward bsz=4, input_seq=256: 0.03526663780212402
Forward bsz=16, input_seq=4: 0.01126551628112793
Forward bsz=16, input_seq=16: 0.012389421463012695
Forward bsz=16, input_seq=256: 0.1560053825378418
Generate benchmarks
==================================================
Generate bsz=1, input_seq=4: 4.527426719665527
Generate bsz=1, input_seq=256: 4.667049169540405
Generate bsz=16, input_seq=4: 5.524803400039673
Generate bsz=16, input_seq=256: 7.931211709976196
```

**Current main:**
```
Forward benchmarks
==================================================
Forward bsz=1, input_seq=4: 0.017528295516967773
Forward bsz=1, input_seq=16: 0.012105464935302734
Forward bsz=1, input_seq=256: 0.01315617561340332
Forward bsz=4, input_seq=4: 0.011912107467651367
Forward bsz=4, input_seq=16: 0.013910531997680664
Forward bsz=4, input_seq=256: 0.035504817962646484
Forward bsz=16, input_seq=4: 0.012083053588867188
Forward bsz=16, input_seq=16: 0.012537956237792969
Forward bsz=16, input_seq=256: 0.15653300285339355
Generate benchmarks
==================================================
Generate bsz=1, input_seq=4: 4.554980516433716
Generate bsz=1, input_seq=256: 4.695344686508179
Generate bsz=16, input_seq=4: 5.55778431892395
Generate bsz=16, input_seq=256: 7.969247102737427
```

=> We don't see any drop in performance at all.

I've verified that the following tests all pass on a single GPU (RTX4090):
FA2:
```
RUN_SLOW=1 pytest -m flash_attn_test tests/models/llama/test_modeling_llama.py
```
and all Llama fast tests:
```
CUDA_VISIBLE_DEVICES="0" RUN_SLOW=1 pytest tests/models/llama/test_modeling_llama.py
```

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -420,12 +466,11 @@ class LlamaFlashAttention2(LlamaAttention):
     def forward(
         self,
         hidden_states: torch.Tensor,
-        attention_mask: Optional[torch.Tensor] = None,
         position_ids: Optional[torch.LongTensor] = None,
         past_key_value: Optional[Tuple[torch.Tensor]] = None,
         output_attentions: bool = False,
         use_cache: bool = False,
-        padding_mask: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.LongTensor] = None,
```
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:473]: It's probably safer to not change the signature ordering.

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -41,6 +41,49 @@
...
+class AttentionMask2DTo4D:
+    def __init__(self, is_causal: bool):
+        self.is_causal = is_causal
+        self.cached_2d_tensor = None
+        self.cached_4d_tensor = None
+
+    def __call__(self, attention_mask_2d: torch.Tensor, input_shape, past_key_values_length, dtype):
+        """
+        Multiplies the given tensor x by -10,000. 
+        If the cached tensor does not exist or has a different size, a new one is allocated.
```
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:53]:
```suggestion
        Prepares and caches an extended 4D float mask of shape (batch_size, 1, tgt_seq_len, src_seq_len), and caches as well attention_mask_2d.
```

**Review by fxmarty — COMMENTED** [review]
Thank you! I am happy with it, just wondering whether changing the `attention_mask` input from being 4D to 2D in `LlamaDecoderLayer` & `LlamaAttention` is considered a breaking change or not.

**Comment by patrickvonplaten** [comment]
> Thank you! I am happy with it, just wondering whether changing the `attention_mask` input from being 4D to 2D in `LlamaDecoderLayer` & `LlamaAttention` is considered a breaking change or not.

To me they are internal classes and thus changing the format of the attention_mask is ok. @LysandreJik @younesbelkada @ArthurZucker what do you think?

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot this looks much cleaner indeed! 
Regarding your comment as those are internal classes I think it is fine, as long as the changes are explicitly detailed in the next release notes

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -41,6 +41,49 @@
 from .configuration_llama import LlamaConfig
 
 
+class AttentionMask2DTo4D:
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:44]: Do we plan to move this to the modelling utils or is this gonna be here for all models?

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:473]: +1 here!

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -870,17 +890,6 @@ def forward(
         if inputs_embeds is None:
             inputs_embeds = self.embed_tokens(input_ids)
         # embed positions
-        if attention_mask is None:
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:873]: super nice to remove this!

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -589,13 +634,13 @@ def _upad_input(self, query_layer, key_layer, value_layer, padding_mask, query_l
 
 
 class LlamaDecoderLayer(nn.Module):
-    def __init__(self, config: LlamaConfig):
+    def __init__(self, config: LlamaConfig, mask_converter=None):
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:637]: If we support passing the mask converter here but not in the parent classes it's kind of pointless no?
Wondering which one would be the best:
1. Pass the mask converter class to all classes
2. Only have it in the attention layer, controlled with a `MASK_CONVERTER = {"default": AttentionMask2DTo4D}` and just in the attention layer do `self.mask_converter = MASK_CONVERTER[config.mask_converter]` with the attribute added to the config common? 
(naming can be improve for sure!)

**Review by ArthurZucker — COMMENTED** [review]
Makes a lot of sense indeed, we have had the past logic for quite a while and an update is welcome!

File: src/transformers/models/llama/modeling_llama.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py:637]: Sorry I don't fully understand this

File: src/transformers/models/llama/modeling_llama.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py:44]: Either #Copied from or we move it to a utils file. Both would work for me

File: src/transformers/models/llama/modeling_llama.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py:646]: To begin with I would not make the Mask Converter a class that the user plays around with, but instead use it as an internal convenience class that doesn't sacrifice speed

Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py]: I see sorry realised that if you want to share the same cached mask you gotta pass it, ignore my comment

File: src/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -538,7 +640,7 @@ def _flash_attention_forward(
                 max_seqlen_k=max_seqlen_in_batch_k,
                 dropout_p=dropout,
                 softmax_scale=softmax_scale,
-                causal=True,
+                causal=self.attention_mask_cache.is_causal,
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:643]: This allows us to easily copy-paste this function to non-causal attention layers (BERT)

File: src/transformers/models/llama/modeling_llama.py [hunk 7]
```diff
@@ -869,22 +948,8 @@ def forward(
 
         if inputs_embeds is None:
             inputs_embeds = self.embed_tokens(input_ids)
-        # embed positions
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:872]: We don't need to unnecessarily create attention mask if not needed. E.g. the 4D attention_mask is never needed for FA2

**Comment by patrickvonplaten** [comment]
This PR should help make the following PRs nicer / cleaner:
- https://github.com/huggingface/transformers/pull/26722
- https://github.com/huggingface/transformers/pull/26572

File: src/transformers/models/llama/modeling_llama.py [hunk 8]
```diff
@@ -609,7 +721,6 @@ def forward(
         past_key_value: Optional[Tuple[torch.Tensor]] = None,
         output_attentions: Optional[bool] = False,
         use_cache: Optional[bool] = False,
-        padding_mask: Optional[torch.LongTensor] = None,
     ) -> Tuple[torch.FloatTensor, Optional[Tuple[torch.FloatTensor, torch.FloatTensor]]]:
         """
         Args:
```
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:726]: @patrickvonplaten Make sure to change the docstring line 728 (of this branch):
```
attention_mask (`torch.FloatTensor`, *optional*): attention mask of size
`(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```

File: src/transformers/models/llama/modeling_llama.py [hunk 9]
```diff
@@ -784,8 +894,14 @@ def __init__(self, config: LlamaConfig):
         self.padding_idx = config.pad_token_id
         self.vocab_size = config.vocab_size
 
+        # create attention mask cache that trickles down to each attention layer
+        # so that the attention_mask cache can be shared among layers
+        attention_mask_cache = AttentionMaskCache(is_causal=True)
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:899]: Re-iterating on the fact that since it's not passed as an argument to the `Model` changing the default class used in the subsequent layers means we have to re-write this class! Let's also support passing it and default to this class wdyt?  
But if we add it in the model then same goes for the XXForCausalLM, which is why choosing here the class form a config could help

**Review by ArthurZucker — APPROVED** [review]
LGTM, we might have to keep some logic to pop the padding mask for 1 release for BC. let's do a deprecation cycle no?

File: src/transformers/models/llama/modeling_llama.py (hunk 9 above)
Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py:899]: I don't follow here really. Do you mean `is_causal=True` should be changed to something like `is_causal=self.causal.is_causal` ?

**Comment by patrickvonplaten** [comment]
> LGTM, we might have to keep some logic to pop the padding mask for 1 release for BC. let's do a deprecation cycle no?

They are all internal classes so BC doesn't really apply here IMO. Also it's only been in one release, so not sure we need to have a deprecation cycle here

File: src/transformers/models/llama/modeling_llama.py (hunk 9 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:899]: No. What I mean is that you are passing the `attention_mask_cache` to the `LlamaDecoderLayer` from the `LlamaModel`, so if anyone wants to use another class (which is part of the reason why we are going with a class design no?) then they are stuck with this one for the  `LlamaModel` . Unless we plan on this only being an internal class and forcing the attention, I think it makes sense to either:
- pass the class or the instance `attention_mask_cache` to the `LlamaModel` (and `LlamaForCausalLM`)
- init the self.attention_mask_cache using a Mapping. 

We know that sliced attention requires a different logic, same for attention sink, so we know for sure that we'll have more attention classes coming no?

**Comment by ArthurZucker** [comment]
Ok, it's true that it has only been 2 weeks, good for me

**Comment by fxmarty** [comment]
@patrickvonplaten I disagree that the non-top-level classes are internal. They are publicly exposed (no `_`), and have docstrings. There are a few libraries that were broken by the addition of padding_mask here: https://github.com/huggingface/transformers/pull/25598

e.g.
https://github.com/casper-hansen/AutoAWQ/pull/88
https://github.com/mosaicml/llm-foundry/pull/643
https://github.com/hpcaitech/ColossalAI/pull/4908

so IMO modifying the non-top level signature / expected args (or args format, e.g. changing an arg from 4D to 2D) classes should be done with great care.

File: src/transformers/models/llama/modeling_llama.py (hunk 9 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:899]: After online talk, agree with you, not probable that the attention mask will change and let's restrict the usage. Could support 4d input that will just be return later, and ok for me to keep it in the modeling file for now

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten I disagree that the non-top-level classes are internal. They are publicly exposed (no `_`), and have docstrings. There are a few libraries that were broken by the addition of padding_mask here: #25598
> 
> e.g. [casper-hansen/AutoAWQ#88](https://github.com/casper-hansen/AutoAWQ/pull/88) [mosaicml/llm-foundry#643](https://github.com/mosaicml/llm-foundry/pull/643) [hpcaitech/ColossalAI#4908](https://github.com/hpcaitech/ColossalAI/pull/4908)
> 
> so IMO modifying the non-top level signature / expected args (or args format, e.g. changing an arg from 4D to 2D) classes should be done with great care.

This PRs all kind of show that no-one understood really how to use the padding_mask if it's just set to None in all the PRs. 

If we force ourselves to deprecate every internal function that doesn't have a `_` prefix (which many people still use in their codebases BTW) , it'll will be very hard to stay agile/dynamic. Ok for me to deprecate (just adding `kwargs` to all forward methods) even though I prefer not to and just state that we correct the `attention_mask` here.

@LysandreJik wdyt?

**Comment by LysandreJik** [comment]
Anything that isn't in the public API is private to us as maintainers, but after spending a bit of time looking in the documentation, I can't find where we have written it (but I remember us writing it down).

In any case the deprecation cycle doesn't cost much and I would keep it for 1-2 versions max before removing it. It was present for very few releases (one) and concerns very few users, so let's do a deprecation cycle with very explicit `v4.37.0` as a cutoff date.

File: src/transformers/models/llama/modeling_llama.py [hunk 10]
```diff
@@ -41,6 +41,143 @@
 from .configuration_llama import LlamaConfig
 
 
+class AttentionMaskCache:
```
Comment (LysandreJik) [inline src/transformers/models/llama/modeling_llama.py:44]: Very cool class! Once porting it to other models, is the idea to have this upstreamed somewhere or would you rather keep it in this class and copy/paste it around?

**Review by LysandreJik — COMMENTED** [review]
The API in itself looks good to me, very clean mask management. Thanks for iterating and benchmarking @patrickvonplaten.

If this class stays very general we can move it to a utility module and put it in the docs.

Happy to have this in the codebase with the slight deprecation mentioned above with some good tests, and to implement it in other models that we want to support

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot @patrickvonplaten !

I can confirm your changes did not introduced any additional overhead for FA2 models, 

Screenshot 2023-10-18 at 10 50 35

Screenshot 2023-10-18 at 10 51 05

Above are the benchmarking results for seq_len=2048 using this script: https://gist.github.com/younesbelkada/02f35734da906cc0f2389ae4f665c58f

File: src/transformers/models/owlv2/modeling_owlv2.py [hunk 11]
```diff
@@ -1378,6 +1378,7 @@ def normalize_grid_corner_coordinates(self, feature_map: torch.FloatTensor):
 
     def objectness_predictor(self, image_features: torch.FloatTensor) -> torch.FloatTensor:
         """Predicts the probability that each image feature token is an object.
+
```
Comment (patrickvonplaten) [inline src/transformers/models/owlv2/modeling_owlv2.py:1381]:
```suggestion
```

File: src/transformers/models/persimmon/modeling_persimmon.py [hunk 12]
```diff
@@ -548,7 +548,6 @@ class PersimmonModel(PersimmonPreTrainedModel):
         config: PersimmonConfig
     """
 
-    # Copied from transformers.models.llama.modeling_llama.LlamaModel.__init__ with LLAMA->PERSIMMON,Llama->Persimmon,PersimmonRMSNorm->nn.LayerNorm,norm->final_layernorm,rms_final_layernorm_eps->layer_norm_eps
```
Comment (patrickvonplaten) [inline src/transformers/models/persimmon/modeling_persimmon.py:551]: IMO removing this is ok as the init is really just 5 lines and I don't think persimmon and Llama are that closely related?! Wdyt @ArthurZucker 

Don't want to add the attn mask converter here as well

File: tests/models/llama/test_modeling_llama.py [hunk 13]
```diff
@@ -39,6 +39,185 @@
         LlamaModel,
         LlamaTokenizer,
     )
+    from transformers.models.llama.modeling_llama import AttentionMaskCache
+
+
+@require_torch
+class AttentionMaskTester(unittest.TestCase):
```
Comment (patrickvonplaten) [inline tests/models/llama/test_modeling_llama.py:46]: I'm testing all the functionalities here

File: src/transformers/models/llama/modeling_llama.py (hunk 10 above)
Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py:67]: I've now added also logic to do the windowed attention mask here as well because windowed attention is allowed to be both 2D in Flash Attention and 4D (cc @ArthurZucker)

Besides causal attn mask, normal attn mask, windowed attention mask, I don't think there are any other masks that are supported by FA2 so this class should be able to handle all use cases.

Mid-term I think we should move this attention mask cache to a general utils as it should be able to handle all use cases. For now, I'd like to leave it here until we've added FA2 to all important models.

File: src/transformers/models/falcon/modeling_falcon.py [hunk 14]
```diff
@@ -88,6 +88,188 @@ def _get_unpad_data(padding_mask):
...
+        mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
+
+        mask = mask.to(dtype)
+
+        if past_key_values_length > 0:
+            mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype, device=device), mask], dim=-1)
+
+        # add lower triangular sliding window mask if necessary
+        if sliding_window is not None:
+            diagonal = past_key_values_length - sliding_window + 1
```
Comment (patrickvonplaten) [inline src/transformers/models/falcon/modeling_falcon.py:252]: @ArthurZucker this corrects a bug that was in Mistral's windowed attention. 

Let's take a look at the following picture:
Screenshot from 2023-10-19 21-29-39

Now let's compute the mask as it was done previously with `-sliding_window`

```py
import torch
print(torch.triu(torch.ones(5,5), diagonal=-3))
```
This gives:

```
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [0., 1., 1., 1., 1.]])
```

which is incorrect.

Doing:
```py
import torch
print(torch.triu(torch.ones(5,5), diagonal=-3+1))
```
gives the correct mask:
```
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [0., 1., 1., 1., 1.],
        [0., 0., 1., 1., 1.]])
```

=> Hence we have a +1 now. Also previously the mask didn't work correctly in case the cache + input > sliding window which could happen if a large prompt is added to a system prompt. This is also corrected here by taking `past_key_values_length` into account.

File: src/transformers/models/llama/modeling_llama.py [hunk 15]
```diff
@@ -63,37 +64,185 @@ def _get_unpad_data(padding_mask):
...
+        mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
+
+        mask = mask.to(dtype)
+
+        if past_key_values_length > 0:
+            mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype, device=device), mask], dim=-1)
+
+        # add lower triangular sliding window mask if necessary
+        if sliding_window is not None:
+            diagonal = past_key_values_length - sliding_window + 1
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:227]: @ArthurZucker this corrects a bug that was in Mistral's windowed attention. 

Let's take a look at the following picture:
Screenshot from 2023-10-19 21-29-39

Now let's compute the mask as it was done previously with `-sliding_window`

```py
import torch
print(torch.triu(torch.ones(5,5), diagonal=-3))
```
This gives:

```
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [0., 1., 1., 1., 1.]])
```

which is incorrect.

Doing:
```py
import torch
print(torch.triu(torch.ones(5,5), diagonal=-3+1))
```
gives the correct mask:
```
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [0., 1., 1., 1., 1.],
        [0., 0., 1., 1., 1.]])
```

=> Hence we have a +1 now. Also previously the mask didn't work correctly in case the cache + input > sliding window which could happen if a large prompt is added to a system prompt. This is also corrected here by taking `past_key_values_length` into account.

**Comment by patrickvonplaten** [comment]
This PR is ready for a review.

I've tested:
All slow tests for Falcon, Mistral, Llama they all pass on GPU.
```
CUDA_VISIBLE_DEVICES="0" RUN_SLOW=1 pytest tests/models/llama/test_modeling_llama.py
```

All flash attn tests:
```
CUDA_VISIBLE_DEVICES="0" RUN_SLOW=1 pytest -m flash_attn_test tests/models
```

There is one slow test that fails for Falcon, but it also fails on "main":
```
FAILED tests/models/falcon/test_modeling_falcon.py::FalconModelTest::test_left_padding_compatibility - AssertionError: False is not true
```
(might also be just my GPU)

As explained here: https://github.com/huggingface/transformers/pull/26792#discussion_r1366021714 this PR also fixes a bug with the sliding window mask for Mistral.


@ArthurZucker @younesbelkada would be great if you could do a final review and run some tests to be sure everything works as expected!

File: src/transformers/models/deprecated/open_llama/modeling_open_llama.py [hunk 16]
```diff
@@ -560,7 +560,7 @@ def get_input_embeddings(self):
     def set_input_embeddings(self, value):
         self.embed_tokens = value
 
-    # Copied from transformers.models.llama.modeling_llama.LlamaModel._prepare_decoder_attention_mask
+    # Copied from transformers.models.bart.modeling_bart.BartDecoder._prepare_decoder_attention_mask
```
Comment (patrickvonplaten) [inline src/transformers/models/deprecated/open_llama/modeling_open_llama.py:563]: Let's copy from original source

**Comment by patrickvonplaten** [comment]
**Update**: I removed all the cache logic and instead just pass the attention_mask in the format that's needed. This is cleaner than caching tensors according to their shape, memory_id, etc...

File: examples/research_projects/jax-projects/big_bird/bigbird_flax.py [hunk 17]
```diff
@@ -9,13 +9,13 @@
 import jax.numpy as jnp
 import joblib
 import optax
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/bigbird_flax.py:11]:
```suggestion
import optax
import wandb
```

File: examples/research_projects/jax-projects/big_bird/train.py [hunk 18]
```diff
@@ -2,11 +2,11 @@
 from dataclasses import replace
 
 import jax
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/train.py:4]:
```suggestion
import jax
import wandb
```

File: examples/research_projects/vqgan-clip/VQGAN_CLIP.py [hunk 19]
```diff
@@ -4,12 +4,12 @@
 import imageio
 import torch
 import torchvision
```
Comment (patrickvonplaten) [inline examples/research_projects/vqgan-clip/VQGAN_CLIP.py:6]:
```suggestion
import torchvision
import wandb
```

File: examples/research_projects/vqgan-clip/VQGAN_CLIP.py [hunk 20]
```diff
@@ -4,12 +4,12 @@
...
-import wandb
 from img_processing import custom_to_pil, loop_post_process, preprocess, preprocess_vqgan
 from loaders import load_vqgan
 from PIL import Image
 from torch import nn
 
+import wandb
```
Comment (patrickvonplaten) [inline examples/research_projects/vqgan-clip/VQGAN_CLIP.py:12]:
```suggestion
```

File: examples/research_projects/jax-projects/big_bird/train.py [hunk 21]
```diff
@@ -2,11 +2,11 @@
...
-import wandb
 from bigbird_flax import Args, DataCollator, FlaxBigBirdForNaturalQuestions, Trainer, build_tx, train_step, val_step
 from datasets import load_dataset
 from flax import jax_utils
 
+import wandb
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/train.py:9]:
```suggestion
```

File: examples/research_projects/jax-projects/big_bird/bigbird_flax.py [hunk 22]
```diff
@@ -9,13 +9,13 @@
...
-import wandb
 from flax import jax_utils, struct, traverse_util
 from flax.serialization import from_bytes, to_bytes
 from flax.training import train_state
 from flax.training.common_utils import shard
 from tqdm.auto import tqdm
 
+import wandb
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/bigbird_flax.py:18]:
```suggestion
```

**Comment by fxmarty** [comment]
Then I am not sure what is the point of the class AttnMaskConverter? By the way, for SDPA, ideally we need both the information of 1/ is padding used 2/ transformers custom attention mask. This is because if custom masking is not used, we may dispatch on flash attention. So passing only a 4D mask for SDPA is suboptimal in my opinion.

Or I could just always pass the 4D attention mask to SDPA, but that kind of defeats the point given that dispatch to FA is then impossible.

**Review by younesbelkada — APPROVED** [review]
I can confirm it looks all good on FA-2 end (benchmarks + tests)! thanks a lot @patrickvonplaten !

File: src/transformers/models/mistral/modeling_mistral.py [hunk 23]
```diff
@@ -332,8 +448,15 @@ def forward(
...
-        padding_mask: Optional[torch.LongTensor] = None,
+        **kwargs,
     ):
+        if "padding_mask" in kwargs:
+            warnings.warn(
+                "Passing `padding_mask` is deprecated and will be removed in v4.37. Please make sure use `attention_mask` instead.`"
+            )
+
+            # overwrite attention_mask with padding_mask
+            attention_mask = kwargs.pop("padding_mask")
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:459]: thanks for this

File: src/transformers/models/llama/modeling_llama.py (hunk 15 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:182]: Good for me!

**Review by ArthurZucker — COMMENTED** [review]
Are we adding the sliding window as a new feature for these models? Otherwise would just use two different classes for Mistral and the other

**Comment by patrickvonplaten** [comment]
> Are we adding the sliding window as a new feature for these models? Otherwise would just use two different classes for Mistral and the other

It's not possible really to use sliding window in Llama because it's hardcoded at initialization "sliding_window=...." for Mistral. So the user can't (and should not use) `sliding_window` for Llama via any config parameters (in the same way `is_causal` is hardcoded to `True`).

=> We do move away a bit from "single-file" policy here as the attention converter is is a general class that has more features that needed for some models. But it does make sense here since there is really not much variation for attention mask across models.

File: src/transformers/models/falcon/modeling_falcon.py [hunk 24]
```diff
@@ -1136,11 +1258,20 @@ def forward(
...
+            key_value_length = seq_length + past_key_values_length
+            # 4d mask is passed through the layers
+            if attention_mask is not None:
+                attention_mask = self.attn_mask_converter.to_4d(
+                    attention_mask, seq_length, key_value_length, dtype=inputs_embeds.dtype
+                )
+            else:
+                attention_mask = self.attn_mask_converter.to_causal_4d(
+                    batch_size, seq_length, key_value_length, dtype=inputs_embeds.dtype, device=inputs_embeds.device
+                )
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/falcon/modeling_falcon.py:1274]: We were also passing a `None` attention mask previously for `padding_mask`

File: src/transformers/models/falcon/modeling_falcon.py [hunk 25]
```diff
@@ -88,6 +89,143 @@ def _get_unpad_data(padding_mask):
...
+        - Create a causal 4d mask with slided window
+        - Convert a 2d attention mask (batch_size, query_length) to a 4d attention mask (batch_size, 1, query_length,
+          key_value_length) that can be multiplied with attention scores
+
+    Parameters:
+        is_causal (`bool`):
+            Whether the attention mask should be a uni-directional (causal) or bi-directional mask.
+
+        sliding_window (`int`, *optional*):
+            Optionally, the sliding window masks can be created if `sliding_window` is defined to a positive integer.
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/falcon/modeling_falcon.py:106]: We loose the copied-from then. I'd expect more Mistral-like models to pop up and think it's not worth removing it, see arguments here: https://github.com/huggingface/transformers/pull/26792#issuecomment-1775354493

**Comment by LysandreJik** [comment]
No problem for me to leave the sliding window in the mask converter class, I indeed think we'll get to see more models leveraging the sliding window (or users that want it supported) in other architectures.
